### PR TITLE
API: Implement new indexing behavior for intervals

### LINKED ISSUE #16316: Confusing (possibly buggy) IntervalIndex behavior
screen shot 2017-05-09 at 10 03 25 pm

In the above, I have a region that I'm querying for with a partially overlapping interval. The query succeeds when the interval is partially overlapping until it doesn't, throwing the key error: 

```
KeyError                                  Traceback (most recent call last)
/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/pandas/core/indexing.py in _has_valid_type(self, key, axis)
   1433                 if not ax.contains(key):
-> 1434                     error()
   1435             except TypeError as e:

/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/pandas/core/indexing.py in error()
   1428                 raise KeyError("the label [%s] is not in the [%s]" %
-> 1429                                (key, self.obj._get_axis_name(axis)))
   1430 

KeyError: 'the label [(5409951, 5409965]] is not in the [index]'
```

I think this is particularly confusing because there doesn't seem to be any prominent difference between the `loc`s that succeed and the `loc` that fails as far as I can tell. I know [we had discussed `loc`'s behavior in this context](https://github.com/pandas-dev/pandas/pull/15309#issuecomment-287936736) but I'm not sure we came to a conclusion. 

By the way, my larger question is about how to find intersections between two `IntervalIndex`. It seems like the [`find_intersections`](https://github.com/pandas-dev/pandas/pull/15309#issuecomment-284081223) function didn't make it into this release @jreback ? Let me know! =]

### LINKED ISSUE #23705: Bug in CategoricalIndex[Interval].__contains__ for non-interval data
```python
In [10]: idx = pd.CategoricalIndex(pd.IntervalIndex.from_breaks([1, 2, 3]))

In [11]: 'a' in idx
```

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-50c294b3bf00> in <module>()
----> 1 'a' in idx

~/sandbox/pandas/pandas/core/indexes/category.py in __contains__(self, key)
    358             return self.hasnans
    359
--> 360         return contains(self, key, container=self._engine)
    361
    362     @Appender(_index_shared_docs['contains'] % _index_doc_kwargs)

~/sandbox/pandas/pandas/core/arrays/categorical.py in contains(cat, key, container)
    200     #  can't be in container either.
    201     try:
--> 202         loc = cat.categories.get_loc(key)
    203     except KeyError:
    204         return False

~/sandbox/pandas/pandas/core/indexes/interval.py in get_loc(self, key, method)
    716                 key = self._maybe_cast_slice_bound(key, 'left', None)
    717
--> 718             start, stop = self._find_non_overlapping_monotonic_bounds(key)
    719
    720             if start is None or stop is None:

~/sandbox/pandas/pandas/core/indexes/interval.py in _find_non_overlapping_monotonic_bounds(self, key)
    664
    665             start = self._searchsorted_monotonic(key, 'left')
--> 666             stop = self._searchsorted_monotonic(key, 'right')
    667         return start, stop
    668

~/sandbox/pandas/pandas/core/indexes/interval.py in _searchsorted_monotonic(self, label, side, exclude_label)
    625             sub_idx = self.left
    626             if self.open_left or exclude_label:
--> 627                 label = _get_prev_label(label)
    628
    629         return sub_idx._searchsorted_monotonic(label, side)

~/sandbox/pandas/pandas/core/indexes/interval.py in _get_prev_label(label)
     90     else:
     91         raise TypeError('cannot determine next label for type {typ!r}'
---> 92                         .format(typ=type(label)))
     93
     94

TypeError: cannot determine next label for type <class 'str'>

```

cc @jschendel

### LINKED ISSUE #24931: IntervalIndex.get_loc doctest failing
```
$ pytest --doctest-modules pandas/core/indexes/interval.py  -k get_loc
========================================================================================================================================================================== test session starts ===========================================================================================================================================================================
platform darwin -- Python 3.7.2, pytest-4.1.0, py-1.6.0, pluggy-0.7.1
rootdir: /Users/taugspurger/sandbox/pandas, inifile: setup.cfg
plugins: xdist-1.23.2, repeat-0.7.0, forked-0.2, faulthandler-1.5.0, cov-2.6.0, hypothesis-3.59.1
collected 12 items / 11 deselected

pandas/core/indexes/interval.py F                                                                                                                                                                                                                                                                                                                                  [100%]

================================================================================================================================================================================ FAILURES ================================================================================================================================================================================
______________________________________________________________________________________________________________________________________________________ [doctest] pandas.core.indexes.interval.IntervalIndex.get_loc ______________________________________________________________________________________________________________________________________________________
735         ---------
736         >>> i1, i2 = pd.Interval(0, 1), pd.Interval(1, 2)
737         >>> index = pd.IntervalIndex([i1, i2])
738         >>> index.get_loc(1)
739         0
740
741         You can also supply an interval or an location for a point inside an
742         interval.
743
744         >>> index.get_loc(pd.Interval(0, 2))
Expected:
    array([0, 1], dtype=int64)
Got:
    slice(0, 2, None)

/Users/taugspurger/sandbox/pandas/pandas/core/indexes/interval.py:744: DocTestFailure
================================================================================================================================================================ 1 failed, 11 deselected in 0.02 seconds =================================================================================================================================================================
```

### LINKED ISSUE #25087: Regression in 0.24: TypeError exception when using dropna on dataframe with categorical index
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

dd = pd.DataFrame(np.arange(10))
dd['x2'] = dd[0] * dd[0]
dd['q'] = pd.qcut(dd['x2'], 5)
dd.set_index('q', inplace=True)
dd.dropna()
```
#### Problem description

The call to dropna raised the following exception: 
```
TypeError: Cannot cast array data from dtype('float64') to dtype('<U32') according to the rule 'safe'
```
This seems to happen only with a categorical index for which the intervals are not all of the same length.  

There was no issue in version 0.23.4 and the issue is not fixed in the master

#### Expected Output
No exception should be raised.
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 25ff4729243d69bada4eaf1eeeebc7ec41418977
python: 3.7.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.19.15-300.fc29.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8
LOCALE: en_CA.UTF-8

pandas: 0.25.0.dev0+44.g25ff47292
pytest: None
pip: 19.0.1
setuptools: 40.6.3
Cython: 0.29.4
numpy: 1.16.1
scipy: 1.2.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: None
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### LINKED ISSUE #25860: [BUG] IntervalIndex.get_loc error for decreasing index
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd                                                     

In [2]: intv1 = pd.IntervalIndex.from_arrays([1, 3, 5], [2, 4, 6])              

In [3]: intv1                                                                   
Out[3]: 
IntervalIndex([(1, 2], (3, 4], (5, 6]],
              closed='right',
              dtype='interval[int64]')

In [4]: intv1.get_loc(intv1[0])                                                 
Out[4]: 0

In [5]: intv2 = pd.IntervalIndex.from_arrays([4, 2, 1], [5, 3, 2])              

In [6]: intv2                                                                   
Out[6]: 
IntervalIndex([(4, 5], (2, 3], (1, 2]],
              closed='right',
              dtype='interval[int64]')

In [7]: intv2.get_loc(intv2[0])                                                 
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-7-00c58bb4fb1e> in <module>
----> 1 intv2.get_loc(intv2[0])

~/Code/pandas/pandas/core/indexes/interval.py in get_loc(self, key, method)
    777                 return slice(start, stop)
    778             else:
--> 779                 raise KeyError(original_key)
    780 
    781         else:

KeyError: Interval(4, 5, closed='right')


```
#### Problem description

This `IntervalIndex` is non-overlapping_monotonic. `IntervalIndex.get_loc` runs in `KeyError`, even the key is its first element. 

After `start, stop = self._find_non_overlapping_monotonic_bounds(key)`, `start` and `stop` get the same value. Thereafter, no conditional handles this case.

#### Expected Outcome

In [5]: intv2 = pd.IntervalIndex.from_arrays([4, 2, 1], [5, 3, 2])              

In [6]: intv2                                                                   
Out[6]: 
IntervalIndex([(4, 5], (2, 3], (1, 2]],
              closed='right',
              dtype='interval[int64]')

In [7]: intv2.get_loc(intv2[0])                                                 
Out[7]: 0

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 01988735f8abde6ddc108f244ef0bbaef35ed1dd
python: 3.7.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.0-16-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_HK.UTF-8
LOCALE: en_HK.UTF-8

pandas: 0.25.0.dev0+308.g01988735f
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #16316, closes #23705, closes #25087, closes #25860, closes #24931
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

I left the `test_interval.py` and `test_interval_new.py` files separate for now, in order to make it more obvious what is being changed.  Will condense these files in a follow-up.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3239,8 +3239,9 @@ def reindex(self, target, method=None, level=None, limit=None,
             if self.equals(target):
                 indexer = None
             else:
-
-                if self.is_unique:
+                # check is_overlapping for IntervalIndex compat
+                if (self.is_unique and
+                        not getattr(self, 'is_overlapping', False)):
```
Comment (jschendel) [inline pandas/core/indexes/base.py]: Kind of gross, but I had to do this in a few places since uniqueness isn't sufficient for `IntervalIndex`, as a query on a unique `IntervalIndex` can still have multiple values returned if they come from the overlap of 2+ intervals.

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -4902,13 +4903,6 @@ def _searchsorted_monotonic(self, label, side='left'):
 
         raise ValueError('index must be monotonic increasing or decreasing')
 
-    def _get_loc_only_exact_matches(self, key):
```
Comment (jschendel) [inline pandas/core/indexes/base.py]: No longer necessary since `get_loc` is now exact for `IntervalIndex`, and was only used with `IntervalIndex`

File: pandas/core/indexes/interval.py [hunk 3]
```diff
@@ -873,109 +849,109 @@ def get_value(self, series, key):
...
-        if self.is_non_overlapping_monotonic:
-            start, stop = self._find_non_overlapping_monotonic_bounds(target)
 
-            start_plus_one = start + 1
-            if not ((start_plus_one < stop).any()):
-                return np.where(start_plus_one == stop, start, -1)
+        if self.is_overlapping:
+            msg = ('cannot handle overlapping indices; use '
+                   'IntervalIndex.get_indexer_non_unique')
+            raise InvalidIndexError(msg)
```
Comment (jschendel) [inline pandas/core/indexes/interval.py]: Changed this to `InvalidIndexError` for consistency with the base method.

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
...
+   Out[5]: True
+
+*New behavior*:
+
+.. ipython:: python
+
+    pd.Interval(1, 2, closed='neither') in ii
+    pd.Interval(-10, 10, closed='both') in ii
+
+The ``get_loc`` method now only returns locations for exact matches to ``Interval`` queries, as opposed to the previous behavior of
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you use ``:meth:`IntervalIndex.get_loc` ``

File: doc/source/whatsnew/v0.25.0.rst [hunk 5]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
...
+   In [6]: ii.get_loc(pd.Interval(1, 5))
+   Out[6]: array([0, 1])
+
+   In [7]: ii.get_loc(pd.Interval(2, 6))
+   Out[7]: array([0, 1, 2])
+
+*New behavior*:
+
+.. code-block:: python
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you use an ipython block and an ok-except?

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
...
+   In [6]: ii.get_loc(pd.Interval(1, 5))
+   Out[6]: 1
+   
+   In [7]: ii.get_loc(pd.Interval(2, 6))
+   ---------------------------------------------------------------------------
+   KeyError: Interval(2, 6, closed='right')
+
+Likewise, ``get_indexer`` and ``get_indexer_non_unique`` will also only return locations for exact matches to ``Interval`` queries, with
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you use a method refernce

File: doc/source/whatsnew/v0.25.0.rst [hunk 7]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
...
 
+.. _whatsnew_0250.api_breaking.interval_indexing:
+
+
+Indexing an ``IntervalIndex`` with ``Interval`` objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Indexing methods for :class:`IntervalIndex` have been modified to return exact matches only for :class:`Interval` queries.
+``IntervalIndex`` methods previously matched on any overlapping ``Interval``.  Behavior with scalar points, e.g. querying
+with an integer, is unchanged (:issue:`16316`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you list any associated issues here

File: doc/source/whatsnew/v0.25.0.rst [hunk 8]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
...
+   In [10]: s.loc[pd.Interval(2, 6)]                                                                                       
+   Out[10]: 
+   (0, 4]    a
+   (1, 5]    b
+   (5, 8]    c
+   dtype: object
+
+*New behavior*:
+
+.. code-block:: python
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: change to an ipython block and use ok-except

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: why do you need the getattr? isn't self.is_overlapping enough?

File: pandas/core/indexes/interval.py [hunk 9]
```diff
@@ -622,6 +623,23 @@ def _maybe_cast_indexed(self, key):
 
         return key
 
+    def _can_reindex(self, indexer):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add types where possible

File: pandas/core/indexes/interval.py [hunk 10]
```diff
@@ -808,58 +814,28 @@ def get_loc(self, key, method=None):
         array([0, 1], dtype=int64)
         """
         self._check_method(method)
+        if is_list_like(key):
+            raise KeyError(key)
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add a comment here on what its doing

File: pandas/core/indexes/interval.py [hunk 11]
```diff
@@ -808,58 +814,28 @@ def get_loc(self, key, method=None):
...
+            op_left = le if self.closed_left else lt
+            op_right = le if self.closed_right else lt
             try:
-                loc = self.get_loc(key)
+                mask = op_left(self.left, key) & op_right(key, self.right)
             except TypeError:
-                # we didn't find exact intervals or are non-unique
-                msg = "unable to slice with this key: {key}".format(key=key)
-                raise ValueError(msg)
+                raise KeyError(key)
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add a commen here on what you are doing

File: pandas/core/indexes/interval.py [hunk 12]
```diff
@@ -873,109 +849,109 @@ def get_value(self, series, key):
...
 
-        if not self.is_unique:
-            raise ValueError("cannot handle non-unique indices")
+        try:
+            target = ensure_index(target)
+        except ValueError:
+            target = Index(target, dtype=object)
 
-        # IntervalIndex
         if isinstance(target, IntervalIndex):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you add some comments on the cases

File: pandas/core/indexes/interval.py [hunk 13]
```diff
@@ -873,109 +849,109 @@ def get_value(self, series, key):
...
-                        indexer.append(np.array([-1]))
-                        continue
+        return ensure_index(indexer), ensure_platform_int(missing)
 
-                # target.closed == 'left'
-                elif self.closed == 'right':
-                    if target_value.left <= value.left:
-                        indexer.append(np.array([-1]))
-                        continue
+    def get_indexer_for(self, target, **kwargs):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: ideally add types where you can

File: pandas/core/indexes/interval.py [hunk 14]
```diff
@@ -873,109 +849,109 @@ def get_value(self, series, key):
...
+        Returns
+        -------
+        numpy.ndarray
+            List of indices.
+        """
+        if self.is_overlapping:
+            return self.get_indexer_non_unique(target, **kwargs)[0]
+        return self.get_indexer(target, **kwargs)
 
+    def get_value(self, series, key):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: does this have a doc-string? (in a superclass?)

**Review by jreback — CHANGES_REQUESTED** [review]
looks really good, some comments.

**Comment by jreback** [comment]
if you have other issues you are closing add them as closes in the top of the PR

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jschendel) [reply pandas/core/indexes/base.py]: It's a safeguard since `is_overlapping` is currently only implemented for `IntervalIndex` and this code can hit any type of index.

It could be added to the base `Index` class but it's slightly ambiguous in terms of meaning - it could just simply return `False` or could essentially be `not self.is_unique` (since duplicate points technically overlap).  Would also contribute to API bloat since it'd be public and not really useful outside of `IntervalIndex` for end users. Depends on what tradeoffs we want to make for internal code clarity.

File: doc/source/whatsnew/v0.25.0.rst (hunk 7 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: Added separate issues under the bug fixes section, as they were more incidental fixes due to indexing being broken in general, and not directly related to the new behavior.

Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: k

File: doc/source/whatsnew/v0.25.0.rst [hunk 15]
```diff
@@ -652,7 +779,8 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
 - Fixed bug in comparison of ordered :class:`Categorical` that contained missing values with a scalar which sometimes incorrectly resulted in ``True`` (:issue:`26504`)
--
+- Bug in :meth:`DataFrame.dropna` when the :class:`DataFrame` has a :class:`CategoricalIndex` containing :class:`Interval` objects incorrectly raised a ``TypeError`` (:issue:`25087`)
+- Bug in :class:`Categorical` and  :class:`CategoricalIndex` with :class:`Interval` values when using the ``in`` operator (``__contains``) with objects that are not comparable to the values in the ``Interval`` (:issue:`23705`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: ``__contains__``

maybe better for these in indexing section?

File: doc/source/whatsnew/v0.25.0.rst [hunk 16]
```diff
@@ -479,6 +479,133 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
 Pandas objects *and* give your subclasses specific ``__str__``/``__repr__`` methods,
 you may have to adjust your ``__str__``/``__repr__`` methods (:issue:`26495`).
 
+.. _whatsnew_0250.api_breaking.interval_indexing:
+
+
+Indexing an ``IntervalIndex`` with ``Interval`` objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Indexing methods for :class:`IntervalIndex` have been modified to return exact matches only for :class:`Interval` queries.
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.25.0.rst]:
```suggestion
Indexing methods for :class:`IntervalIndex` have been modified to require exact matches only for :class:`Interval` queries.
```

File: doc/source/whatsnew/v0.25.0.rst (hunk 5 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.25.0.rst]: I would keep the code-block, as otherwise you get a long error traceback which is not that nice to show in the docs

File: doc/source/whatsnew/v0.25.0.rst (hunk 8 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.25.0.rst]: Same as above

File: pandas/core/indexes/interval.py [hunk 17]
```diff
@@ -791,75 +797,46 @@ def get_loc(self, key, method=None):
...
         >>> i3 = pd.Interval(0, 2)
-        >>> overlapping_index = pd.IntervalIndex([i2, i3])
-        >>> overlapping_index.get_loc(1.5)
-        array([0, 1], dtype=int64)
+        >>> overlapping_index = pd.IntervalIndex([i1, i2, i3])
+        >>> overlapping_index.get_loc(0.5)
+        array([ True, False,  True])
+
+        Only exact matches will be returned if an interval is provided.
+        >>> index.get_loc(pd.Interval(0, 1))
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/interval.py]: blank line above this one

File: doc/source/whatsnew/v0.25.0.rst (hunk 4 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: doc/source/whatsnew/v0.25.0.rst (hunk 5 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: doc/source/whatsnew/v0.25.0.rst (hunk 6 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: doc/source/whatsnew/v0.25.0.rst (hunk 8 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: pandas/core/indexes/interval.py (hunk 9 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 10 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 11 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 12 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 13 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: done

File: pandas/core/indexes/interval.py (hunk 14 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: created a shared docstring from the superclass

File: doc/source/whatsnew/v0.25.0.rst (hunk 15 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: pandas/tests/indexes/interval/test_interval_new.py [hunk 18]
```diff
@@ -127,41 +126,46 @@ def test_slice_locs_with_ints_and_floats_succeeds(self):
...
+        ([Interval(2, 4, closed='right'), Interval(2, 4, closed='right')],
+         [1, 1]),
+        ([Interval(5, 7, closed='right'), Interval(2, 4, closed='right')],
+         [2, 1]),
+        ([Interval(2, 4, closed='right'), Interval(2, 4, closed='left')],
          [1, -1])])
     def test_get_indexer_with_interval(self, query, expected):
 
-        tuples = [(0, 2.5), (1, 3), (2, 4)]
+        tuples = [(0, 2), (2, 4), (5, 7)]
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/interval/test_interval_new.py]: change here was to have a non-overlapping index?

File: pandas/tests/indexing/interval/test_interval_new.py [hunk 19]
```diff
@@ -107,22 +104,22 @@ def test_loc_with_slices(self):
...
 
-        expected = s.iloc[4:]
+        expected = s.iloc[3:]
         result = s.loc[Interval(3, 4):]
         tm.assert_series_equal(expected, result)
         result = s[Interval(3, 4):]
         tm.assert_series_equal(expected, result)
 
-        with pytest.raises(KeyError):
+        with pytest.raises(NotImplementedError):
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/interval/test_interval_new.py]: It feels a bit strange to return a NotImplementedError here. I would at least add a error message about slicing not being implemented for non-exact matches

**Comment by jorisvandenbossche** [comment]
It would also be good to add some documentation on indexing in the advanced.rst interval section (maybe copy some example from the whatsnew you added)

File: pandas/tests/indexes/interval/test_interval_new.py (hunk 18 above)
Comment (jschendel) [reply pandas/tests/indexes/interval/test_interval_new.py]: Yes, since overlapping indexes are invalid for `get_indexer` and need to use `get_indexer_non_unique` instead.  For intervals, overlapping is the analogue for uniqueness, as uniqueness isn't a strong enough condition since a query on a unique `IntervalIndex` can still have multiple values returned if they come from the overlap of 2+ intervals.

File: pandas/tests/indexing/interval/test_interval_new.py (hunk 19 above)
Comment (jschendel) [reply pandas/tests/indexing/interval/test_interval_new.py]: Done. Since the `NotImplementedError` comes from `_searchsorted_monotonic` I had to make the message a bit more generic, as it could conceivably be raised in other scenarios.

File: doc/source/whatsnew/v0.25.0.rst (hunk 8 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.25.0.rst]: reverted to the python block per @jorisvandenbossche's comments

**Comment by jschendel** [comment]
@jorisvandenbossche : added a description of the new behavior to `advanced.rst`.

File: doc/source/user_guide/advanced.rst [hunk 20]
```diff
@@ -965,6 +965,28 @@ If you select a label *contained* within an interval, this will also select the
    df.loc[2.5]
    df.loc[[2.5, 3.5]]
 
+Selecting using an ``Interval`` will only return exact matches.
```
Comment (jorisvandenbossche) [inline doc/source/user_guide/advanced.rst]: I would maybe put this (together we the above examples) in a subsection on indexing with an IntervalIndex

**Review by jorisvandenbossche — COMMENTED** [review]
Updates look good!

File: doc/source/user_guide/advanced.rst [hunk 21]
```diff
@@ -965,6 +965,28 @@ If you select a label *contained* within an interval, this will also select the
...
+   ---------------------------------------------------------------------------
+   KeyError: Interval(0.5, 2.5, closed='right')
+
+Selecting all ``Intervals`` that overlap a given ``Interval`` can be performed using the
+:meth:`~IntervalIndex.overlaps` method to create a boolean indexer.
+
+.. ipython:: python
+
+   idxr = df.index.overlaps(pd.Interval(0.5, 2.5))
+   df[idxr]
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst]: show idxr as well

File: doc/source/user_guide/advanced.rst (hunk 20 above)
Comment (jschendel) [reply doc/source/user_guide/advanced.rst]: Created a section for indexing, and also one for binning since the `cut` examples seemed out of place otherwise.

File: doc/source/user_guide/advanced.rst (hunk 21 above)
Comment (jschendel) [reply doc/source/user_guide/advanced.rst]: done (did the same in the whatsnew too)

**Comment by jschendel** [comment]
@jreback @jorisvandenbossche : also removed the warning about indexing behavior being provisional

**Comment by jschendel** [comment]
@jreback @jorisvandenbossche : all existing comments have been addressed

File: doc/source/user_guide/advanced.rst [hunk 22]
```diff
@@ -965,6 +964,32 @@ If you select a label *contained* within an interval, this will also select the
    df.loc[2.5]
    df.loc[[2.5, 3.5]]
 
+Selecting using an ``Interval`` will only return exact matches.
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst]: can you add a versionchanged 0.25 here

File: doc/source/whatsnew/v0.25.0.rst [hunk 23]
```diff
@@ -484,6 +484,142 @@ This change is backward compatible for direct usage of Pandas, but if you subcla
 Pandas objects *and* give your subclasses specific ``__str__``/``__repr__`` methods,
 you may have to adjust your ``__str__``/``__repr__`` methods (:issue:`26495`).
 
+.. _whatsnew_0250.api_breaking.interval_indexing:
+
+
+Indexing an ``IntervalIndex`` with ``Interval`` objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Indexing methods for :class:`IntervalIndex` have been modified to require exact matches only for :class:`Interval` queries.
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you add a link the docs you added above in indexing

File: pandas/core/indexes/interval.py [hunk 24]
```diff
@@ -870,112 +859,140 @@ def get_value(self, series, key):
...
-
-            start_plus_one = start + 1
-            if not ((start_plus_one < stop).any()):
-                return np.where(start_plus_one == stop, start, -1)
-
-        if not self.is_unique:
-            raise ValueError("cannot handle non-unique indices")
+        try:
+            target = ensure_index(target)
+        except ValueError:
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: hmm, why is this needed? ensure_index should just work here

File: pandas/core/indexes/interval.py [hunk 25]
```diff
@@ -870,112 +859,140 @@ def get_value(self, series, key):
...
-            elif rhs == -1:
+    @Appender(_index_shared_docs['get_indexer_non_unique'] % _index_doc_kwargs)
+    def get_indexer_non_unique(self,
+                               target: AnyArrayLike
+                               ) -> Tuple[np.ndarray, np.ndarray]:
 
-                indexer.append(np.arange(lhs, n))
+        try:
+            target = ensure_index(target)
+        except ValueError:
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: same question here

File: pandas/core/indexes/interval.py [hunk 26]
```diff
@@ -870,112 +859,140 @@ def get_value(self, series, key):
...
+            target = Index(target, dtype=object)
 
-            elif lhs == -1:
+        # check that target IntervalIndex is compatible
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: maybe worth it to make a helper method for indexing with an II (as this is the same code as in get_indexer for unique)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also merge master

File: pandas/core/indexes/interval.py (hunk 24 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: Looks like it `ensure_index` raises when provided a list of mixed closed intervals:
```python-traceback
In [6]: ivs = [pd.Interval(0, 1, closed='left'), 
   ...:        pd.Interval(1, 2, closed='right')]

In [7]: ensure_index(ivs)
---------------------------------------------------------------------------
ValueError: intervals must all be closed on the same side
```
I'm guessing we'd want `ensure_index` to have this return an `object` dtype `Index`?  Would we also want `Index(ivs)` to do the same?

Comment (jreback) [reply pandas/core/indexes/interval.py]: yep i think it guarantees to return an Index (which maybe object)

Comment (jschendel) [reply pandas/core/indexes/interval.py]: Okay, found a fix; it's pretty small but will do it in a separate PR later tonight, as the scope of this PR is large enough already (unless you'd rather it be done here).

Comment (jreback) [reply pandas/core/indexes/interval.py]: can u do as a precursor ?

**Comment by jreback** [comment]
merge master

File: pandas/core/indexes/interval.py (hunk 24 above)
Comment (jreback) [reply pandas/core/indexes/interval.py]: this can now be removed (and below)

**Review by jreback — CHANGES_REQUESTED** [review]
2 small changes that remove code done in the precursor and i think good to go

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jschendel very nice patch!

**Comment by zfrenchee** [comment]
Glad to see this finally resolved. Looking forward to trying out the new behavior.

**Comment by jorisvandenbossche** [comment]
@zfrenchee it would be very welcome to test and give feedback ! 
(there should also be a release candidate in a few days, if that is easier to test than master)

**Comment by alexlenail** [comment]
@jorisvandenbossche I noticed today 

```
NotImplementedError: contains not implemented for two intervals
```

Both in 25.3 and master. Is there still room to contribute to intervalIndex?

**Comment by jorisvandenbossche** [comment]
@alexlenail I think it is best to open a new issue for this (to ensure it doesn't get lost) 
I need to look back whether this was done on purpose or just because it was less straightforward / lack of time (the error type seems to suggest the second). But let's use a new issue to discuss that. Contributions are certainly welcome!
